src/Agent.Worker/Release/Artifacts/FileContainer/DeployConnectionFactory.cs

```diff
@@ -0,0 +1,91 @@
+﻿using System;
+using System.Collections.Concurrent;
+using System.Net.Http;
+using System.Threading.Tasks;
+using Microsoft.VisualStudio.Services.Agent.Worker;
+using Microsoft.VisualStudio.Services.Client;
+using Microsoft.VisualStudio.Services.Common;
+using Microsoft.VisualStudio.Services.OAuth;
+using Microsoft.VisualStudio.Services.WebApi;
+
+namespace Agent.Worker.Release.Artifacts.FileContainer
+{
+    /// <summary>
+    /// Our factory for VssConnections and TfsConnections. Used to ensure:
+    ///  1. consistent initialization
+    ///  2. No credential dialogs if NoCredentialPrompt is set.
+    ///  3. Connect() is called, but only once for a given Uri.
+    /// </summary>
+    public static class DeployConnectionFactory
+    {
+        public static async Task<VssConnection> GetVssConnectionAsync(Uri uri, string accessToken, IExecutionContext logger = null)
+        {
+            VssConnection connection;
+            if (!s_connections.TryGetValue(uri, out connection))
+            {
+                VssClientCredentials cred = GetCredentials(accessToken);
+
+                VssClientHttpRequestSettings settings = VssClientHttpRequestSettings.Default.Clone();
+
+                if (settings.SendTimeout < s_minTimeout)
+                {
+                    settings.SendTimeout = s_minTimeout;
+                }
+
+                if (settings.MaxRetryRequest < s_minRetries)
+                {
+                    settings.MaxRetryRequest = s_minRetries;
+                }
+
+                VssHttpMessageHandler innerHandler = new VssHttpMessageHandler(cred, settings);
+
+                // Retry on timeouts too.
+                HttpRetryOnTimeoutOptions timeoutRetryOptions = new HttpRetryOnTimeoutOptions()
+                {
+                    MaxRetries = s_minRetries,
+                    MinBackoff = s_minBackoff,
+                    BackoffCoefficient = s_backoffCoefficient,
+                };
+
+                DelegatingHandler[] handlers = new DelegatingHandler[]
+                {
+                    new HttpRetryOnTimeoutMessageHandler(timeoutRetryOptions, logger),
+                };
+
+                connection = new VssConnection(uri, innerHandler, handlers);
+                await connection.ConnectAsync().ConfigureAwait(false);
```

Comment:
no need to maintain Synchronization context, the continuation can be run by a different thread.